fedimint-client/src/meta.rs

```diff
@@ -0,0 +1,262 @@
+            initial_fetch_waiter: Waiter::new(),
+            source,
+        })
+    }
+
+    /// Get the value for the meta field.
+    ///
+    /// This may wait for significant time on first run.
+    pub async fn get_field<V: DeserializeOwned + 'static>(
+        &self,
+        db: &Database,
+        field: &str,
+    ) -> Option<TypedMetaValue<V>> {
+        if let Some(value) = self.get_field_from_db(db, field).await {
+            // might be from in old cache.
+            // TODO: maybe old cache should have a ttl?
+            Some(value)
+        } else {
+            // wait for initial value
+            self.initial_fetch_waiter.wait().await;
+            self.get_field_from_db(db, field).await
+        }
+    }
+
+    async fn get_field_from_db<V: DeserializeOwned + 'static>(
+        &self,
+        db: &Database,
+        field: &str,
+    ) -> Option<TypedMetaValue<V>> {
+        db.begin_transaction_nc()
+            .await
+            .get_value(&MetaFieldKey(field.to_string()))
+            .await
+            .and_then(|value| {
+                Some(TypedMetaValue {
+                    revision: value.revision,
+                    value: parse_meta_value_static::<V>(&value.value).ok()?,
+                })
+            })
+    }
+
+    /// Update all source in background.
+    ///
+    /// Caller should run this method in a task.
+    pub(crate) async fn update_continuously(&self, client: &Client) -> ! {
+        let meta = self.source.fetch(client, FetchKind::Initial).await;
+        let failed_initial = meta.is_err();
+        match meta {
+            Ok(meta) => Self::save_meta_values(client, &meta).await,
+            Err(error) => warn!(%error, "failed to fetch source"),
+        };
+        self.initial_fetch_waiter.done();
+
+        // now keep updating slowly
+        loop {
+            // don't wait if we failed first item
+            if !failed_initial {
+                MetaSource::wait_for_update(&self.source).await;
+            }
+
+            if let Ok(meta) = self.source.fetch(client, FetchKind::Background).await {
+                Self::save_meta_values(client, &meta).await;
+            }
+        }
+    }
+
+    async fn save_meta_values(client: &Client, meta: &BTreeMap<MetaFieldKey, MetaFieldValue>) {
+        let mut dbtx = client.db().begin_transaction().await;
+        dbtx.remove_by_prefix(&MetaFieldPrefix).await;
+        for (key, value) in meta {
+            dbtx.insert_entry(key, value).await;
+        }
+        dbtx.commit_tx().await;
+    }
+}
+
+/// Legacy non-meta module config source uses client config meta and
+/// meta_override_url meta field.
+#[derive(Clone, Debug, Default)]
+#[non_exhaustive]
+pub struct LegacyMetaSource {
+    reqwest: reqwest::Client,
+}
+
+#[apply(async_trait_maybe_send!)]
+impl MetaSource for LegacyMetaSource {
+    async fn wait_for_update(&self) {
+        fedimint_core::task::sleep(Duration::from_secs(10 * 60)).await
+    }
+
+    async fn fetch(
```

Comment:
fetch, should take `existing_revision: Option<u64>`, so in case of meta module, it can issue `check-revision` and 99.99% avoid downloading everything again.

Reply:
revision is per field right?